src/Schema.ts

```diff
@@ -88,10 +87,66 @@ interface DictionaryTypeDetail {
 	dictionary: string;
 	dataTransformer: DataTransformer;
 }
+
+/** Infer whether a schema type definition is required and union the result with null if it is not */
+type InferRequiredType<TScalar, TType> = TScalar extends { required: true } ? TType : TType | null;
+
+/**
+ * Infer the output of a string type definition with specific handling for enumerations
+ * If an enumeration is a readonly array, the return type of the definition will be a union
+ * of the array elements. Otherwise, the return type will be a string.
+ */
+type InferStringType<TString extends SchemaTypeDefinitionString> =
+	TString['enum'] extends readonly (infer E)[] ? E : string;
+
+/** Infer the output type of a schema type definition */
+type InferSchemaType<TSchemaTypeDefinition> =
+	TSchemaTypeDefinition extends SchemaTypeDefinitionBoolean
+		? InferRequiredType<TSchemaTypeDefinition, boolean>
+		: TSchemaTypeDefinition extends SchemaTypeDefinitionString
+			? InferRequiredType<TSchemaTypeDefinition, InferStringType<TSchemaTypeDefinition>>
+			: TSchemaTypeDefinition extends SchemaTypeDefinitionNumber
+				? InferRequiredType<TSchemaTypeDefinition, number>
+				: TSchemaTypeDefinition extends SchemaTypeDefinitionISOCalendarDate
+					? InferRequiredType<TSchemaTypeDefinition, `${number}-${number}-${number}`>
+					: TSchemaTypeDefinition extends SchemaTypeDefinitionISOCalendarDateTime
+						? InferRequiredType<
+								TSchemaTypeDefinition,
+								`${number}-${number}-${number}T${number}:${number}:${number}.${number}`
+							>
+						: TSchemaTypeDefinition extends SchemaTypeDefinitionISOTime
+							? InferRequiredType<TSchemaTypeDefinition, `${number}:${number}:${number}.${number}`>
+							: TSchemaTypeDefinition extends Schema<infer SubSchemaDefinition>
+								? InferSchemaType<SubSchemaDefinition>
+								: TSchemaTypeDefinition extends SchemaTypeDefinitionArray
+									? InferSchemaType<TSchemaTypeDefinition[0]>[]
+									: TSchemaTypeDefinition extends SchemaDefinition
+										? {
+												[K in keyof TSchemaTypeDefinition]: InferSchemaType<
+													TSchemaTypeDefinition[K]
+												>;
+											}
+										: never;
+
+/** Infer the shape of a `Document` instance based upon the Schema it was instantiated with */
+export type InferDocumentObject<TSchema extends Schema<SchemaDefinition>> =
+	TSchema extends Schema<infer TSchemaDefinition>
+		? {
+				[K in keyof TSchemaDefinition]: InferSchemaType<TSchemaDefinition[K]>;
+			}
+		: never;
+
+/** Infer the shape of a `Model` instance based upon the Schema it was instantiated with */
+export type InferModelObject<TSchema extends Schema<SchemaDefinition>> = {
+	_id: string;
+	__v: string;
+} & InferDocumentObject<TSchema> extends infer O
```

Comment:
Does it make sense to have a special Model type for `_raw` records?

Since there's no Schema how dos this inference work?

Reply:
I made some changes to improve the experience of schema-less models and documents in a794f2bbb09a806bae03d00e13fadf04d0d92766. It wasn't really broken previously -- it just ended up using a generic `SchemaDefinition` which couldn't be used to infer anything at all. However, it could be improved, particularly around the behavior of the `_raw` property to ensure it is typed appropriately based on whether a schema is supplied or not.

I further added tests in 0a3cb8d41cda648085a030c7e379ba1ee2a8d443 to confirm the output types for `Document` and `Model` methods and included both schema-less and schema variants in them.